### PR TITLE
WT-13377 Fix session->reconfigure with respect to cache_max_wait_ms, accept 0

### PR DESCRIPTION
**Pull request by luke-pearson**
1. Fixed a bug that meant callers could not set cache_max_wait_ms to 0.
2. Implemented a unit test for this functionality.
3. Extended it to all but the prefetch configuration logic which is too special for easy unit testing.

### CODE REVIEW

File: src/session/session_api.c
Comment (tod-johnson-mongodb) [inline src/session/session_api.c:516]: The original code tests config in the order ignore_cache_size, cache_cursors, debug.checkpoint_fail_before_turtle_update, debug.release_evict_page, and cache_max_wait_ms. The new __session_config_int() tests in the a different order: ignore_cache_size, cache_cursors, debug.checkpoint_fail_before_turtle_update, cache_max_wait_ms, and debug.release_evict_page. If there are any errors other than WT_NOTFOUND the result is different. Otherwise it does not mater. In any case, is there a reason to change the order?

File: test/unittest/tests/test_session_config.cpp
Comment (tod-johnson-mongodb) [inline test/unittest/tests/test_session_config.cpp:28]: I believe new line 28 is intended to test that __session_config_int() returns 0 if config does not match any of the config parameters it parses.I suggest to add another test that mentions a parameter not parsed by __session_config_int(). Perhaps add
REQUIRE(__ut_session_config_int(session, "not_a_real_parameter=false") == 0);

File: test/unittest/tests/test_session_config.cpp
Comment (tod-johnson-mongodb) [inline test/unittest/tests/test_session_config.cpp:52]: I suggest changing "with invalid strings" to "with invalid strings or valid strings for parameters not parsed by it."
Then add one more test:
                REQUIRE(__ut_session_config_int(session, "not_a_real_parameter") == 0);

File: test/unittest/tests/test_session_config.cpp
Comment (tod-johnson-mongodb) [inline test/unittest/tests/test_session_config.cpp:8]: Add a comment at the start of the file stating what the test is testing.

**Review by tod-johnson-mongodb — APPROVED** [review]
The code is correct as written. I give a few improvements. The most important is a comment at the beginning of the file stating what is being tested. I would also like to see comments explaining the individual tests, but we haven't agreed how to do that yet.

Another important one is the order of some checks were changed without cause.

File: src/session/session_api.c
Comment (ajmorton) [inline src/session/session_api.c:2761]: This probably falls under the Q3 testing scope, but is there a way we can expose these functions to unit testing without writing one wrapper per function? 
I suspect the end result of this solution will be one wrapper function for every static function in the code base.

File: test/unittest/tests/test_session_config.cpp
Comment (ajmorton) [inline test/unittest/tests/test_session_config.cpp:17]:
```suggestion
#define BOOLEAN_FLAG_TEST(config_param, flag)                                                    \
```

File: test/unittest/tests/test_session_config.cpp
Comment (ajmorton) [inline test/unittest/tests/test_session_config.cpp:38]: How did you decide on this set of config items to test? 
Is it `all boolean flags` + `cache_max_wait_ms`, and the rest will come in a later ticket?

File: test/unittest/tests/test_session_config.cpp
Comment (ajmorton) [inline test/unittest/tests/test_session_config.cpp:23]: We can also test for invalid arguments here. The [csuite config test](https://github.com/wiredtiger/wiredtiger/blob/develop/test/csuite/config/main.c) has some examples we can reference.

**Review by ajmorton — APPROVED** [review]
LGTM pending Tod's comments.
I've left a few questions/suggestions, but nothing that _needs_ to happen before merge.

File: test/unittest/tests/test_session_config.cpp
Comment (jeremythorp) [inline test/unittest/tests/test_session_config.cpp:17]: Catch2's [SECTION](https://github.com/catchorg/Catch2/blob/devel/docs/test-cases-and-sections.md) is a great way of running a set of tests which share common setup & teardown code.

For example, FLAG_TEST could be replaced with something like:

`static void
test_config_flag(WT_SESSION_IMPL *session, const std::string& config_param, const int flag)
{
    std::string param_true = config_param + "=true";
    std::string param_false = config_param + "=false";
    REQUIRE(__ut_session_config_int(session, param_true.c_str()) == 0);
    REQUIRE(F_ISSET(session, flag));
    REQUIRE(__ut_session_config_int(session, param_false.c_str()) == 0);
    REQUIRE(!F_ISSET(session, flag));
    REQUIRE(__ut_session_config_int(session, param_true.c_str()) == 0);
    REQUIRE(__ut_session_config_int(session, "") == 0);
    REQUIRE(F_ISSET(session, flag));
}

TEST_CASE("Session config - test setting and clearing a flag")
{
    /* Build Mock session, this will automatically create a mock connection. */
    std::shared_ptr<MockSession> session_mock = MockSession::buildTestMockSession();
    WT_SESSION_IMPL *session = session_mock->getWtSessionImpl();

    SECTION("ignore_cache_size")
    {
        test_config_flag(session, "ignore_cache_size", WT_SESSION_IGNORE_CACHE_SIZE);
    }

    SECTION("cache_cursors")
    {
        test_config_flag(session, "cache_cursors", WT_SESSION_CACHE_CURSORS);
    }

    SECTION("debug.checkpoint_fail_before_turtle_update")
    {
        test_config_flag(session, "debug.checkpoint_fail_before_turtle_update", WT_SESSION_DEBUG_CHECKPOINT_FAIL_BEFORE_TURTLE_UPDATE);
    }

    SECTION("debug.release_evict_page")
    {
        test_config_flag(session, "debug.release_evict_page", WT_SESSION_DEBUG_RELEASE_EVICT);
    }
}`

File: test/unittest/tests/test_session_config.cpp
Comment (luke-pearson) [reply test/unittest/tests/test_session_config.cpp:8]: 838a57d7b

File: test/unittest/tests/test_session_config.cpp
Comment (luke-pearson) [reply test/unittest/tests/test_session_config.cpp]: Jeremy found a better solution, going with that.

Comment (luke-pearson) [reply test/unittest/tests/test_session_config.cpp]: 838a57d7b

File: src/session/session_api.c
Comment (luke-pearson) [reply src/session/session_api.c:2761]: cc: @korteland, @jeremythorp. Andrew is correct that at some point we will have hundreds of these wrapper. I don't think there is a way around them. Are any other methods possible?

File: test/unittest/tests/test_session_config.cpp
Comment (luke-pearson) [reply test/unittest/tests/test_session_config.cpp]: This is the only set of config items that can be pulled into it's own configuration function. The decision process was:

1. Test cache_max_wait_ms, okay that was relatively easy. May as well extend it
2. Try and extend it to all of the things explicitly set in session_reconfigure. So not things that get set in __wt_txn_reconfigure. 
3. Anything that does weird things like prefetch is out of scope.

File: test/unittest/tests/test_session_config.cpp
Comment (luke-pearson) [reply test/unittest/tests/test_session_config.cpp]: I think at that point we're testing __wt_config_getones and not the function itself which is why I've avoided this.

File: src/session/session_api.c
Comment (luke-pearson) [reply src/session/session_api.c:516]: 2f051ad88

Comment (tod-johnson-mongodb) [reply src/session/session_api.c:516]: Thank you.

File: test/unittest/tests/test_session_config.cpp
Comment (tod-johnson-mongodb) [reply test/unittest/tests/test_session_config.cpp:78]: Thank you.

File: test/unittest/tests/test_session_config.cpp
Comment (tod-johnson-mongodb) [reply test/unittest/tests/test_session_config.cpp:8]: Thank you.

File: src/session/session_api.c
Comment (tod-johnson-mongodb) [reply src/session/session_api.c:2761]: An idea: for every file file.c that contains static functions have a file file_wrappers.h that contains all the static function wrappers. File file.c would include file_wrappers.h at the end. This way the wrappers do not get in the way while reading file file.c.

Comment (ajmorton) [reply src/session/session_api.c:2761]: We could also look at a `WT_STATIC` macro which behaves like `WT_INLINE` and is removed by the pre-processer when we compile with a certain flag. I think we've discussed this option in the past but I don't remember why we didn't pursue it.

**Comment by luke-pearson** [comment]
- [x] #739
- [ ] https://github.com/octo-org/octo-repo/issues/740
- [ ] Add delight to the experience when all tasks are complete :tada:

File: src/session/session_api.c
Comment (jeremythorp) [reply src/session/session_api.c:2761]: @korteland and I introduced the __ut_ wrapper functions as a minimal, if not perfect, change to expose these functions for unit testing. I'm more than happy to consider other options that are cleaner but required more changes (such as splitting the current headers into multiple more focussed ones). The key thing is that internal functions need to be exposed for unit/sub-API testing.

File: test/unittest/tests/test_session_config.cpp
Comment (jeremythorp) [reply test/unittest/tests/test_session_config.cpp]: Thanks!

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Pull request merged**
